File: repo.py
import time
from pyndn import Name
from pyndn import Data
from pyndn import Face
from pyndn import Interest
from pyndn.security import KeyChain
from pyndn.meta_info import MetaInfo

def dump(*list):
    result = ""
    for element in list:
        result += (element if type(element) is str else str(element)) + " "
    print(result)

metaInfo = MetaInfo()
metaInfo.setFreshnessPeriod(30)

savedMeta = None

class Echo(object):
    def __init__(self, keyChain, certificateName):
        self._keyChain = keyChain
        self._certificateName = certificateName
        self._responseCount = 0

    def onInterest(self, prefix, interest, face, interestFilterId, filter):
        global savedMeta
        self._responseCount += 1

        if interest.getName()[-1].toEscapedString() == '_meta' :
            dump("NEEDED ", interest.getName())
            if savedMeta:
                dump(" > REPLY ", savedMeta.getName())
                face.putData(savedMeta)
            else:
                dump(" > STORAGE EMPTY")
        elif interest.getName()[-1].toEscapedString() == '_latest' :
            dump("NEEDED ", interest.getName())
            dump(" > IGNORE")

    def onRegisterFailed(self, prefix):
        self._responseCount += 1
        dump("Register failed for prefix", prefix.toUri())

def main():
    face = Face()
    keyChain = KeyChain()
    face.setCommandSigningInfo(keyChain, keyChain.getDefaultCertificateName())

    echo = Echo(keyChain, keyChain.getDefaultCertificateName())
    prefix = Name("/ndn/repo/case/test")
    dump("Register prefix", prefix.toUri())
    face.registerPrefix(prefix, echo.onInterest, echo.onRegisterFailed)

    def fetchMeta():
        i = Interest(Name(prefix).append('_meta'))
        i.setMustBeFresh(True)
        i.setInterestLifetimeMilliseconds(3000)

        def onData(i, d):
            global savedMeta
            savedMeta = d 
            dump(' > SAVED', d.getName(), d.getContent().toRawStr())

        dump('REQUEST', i.getName())
        face.expressInterest(i, onData)

    metaFetchInterval = 1/2
    timestampCheck = { 'meta_fetch' : time.time() }

    while True:
        now = time.time()
        if now - timestampCheck['meta_fetch'] >= metaFetchInterval:
            timestampCheck['meta_fetch'] = now
            fetchMeta()

        face.processEvents()
        time.sleep(0.01)

    face.shutdown()

main()

File: cpp/repo.cpp
// g++ -o repo repo.cpp -std=c++11 -I../../ndnrtc-thirdparty/ndn-cpp/build/include -L../../ndnrtc-thirdparty/ndn-cpp/build/lib -lndn-cpp
#include <iostream>
#include <chrono>
#include <ndn-cpp/name.hpp>
#include <ndn-cpp/data.hpp>
#include <ndn-cpp/interest.hpp>
#include <ndn-cpp/security/key-chain.hpp>
#include <ndn-cpp/util/memory-content-cache.hpp>

using namespace std;
using namespace std::chrono;
using namespace ndn;

int main()
{
  Face face = Face();
  KeyChain keyChain = KeyChain();
  face.setCommandSigningInfo(keyChain, keyChain.getDefaultCertificateName());
  Name prefix("/ndn/repo/case/test");
  ptr_lib::shared_ptr<Data> savedMeta;

  face.registerPrefix(prefix,
    [&savedMeta](const ptr_lib::shared_ptr<const Name>& prefix,
       const ptr_lib::shared_ptr<const Interest>& interest, Face& face,
       uint64_t interestFilterId,
       const ptr_lib::shared_ptr<const InterestFilter>& filter)
      {
         if (interest->getName()[-1].toEscapedString() == "_meta")
         {
           cout << "NEEDED " << interest->getName() << endl;
           if (savedMeta)
           {
             cout << " > REPLY " << savedMeta->getName() << endl;
             face.putData(*savedMeta);
           }
           else
            cout << " > STORAGE EMPTY" << endl;
         }
       },
    [](const ptr_lib::shared_ptr<const Name>& prefix){
      cout << "register failed for prefix " << prefix << endl;
    });

    auto fetchMeta = [prefix, &savedMeta, &face](){
      Interest i(Name(prefix).append("_meta"));
      i.setMustBeFresh(true);
      i.setInterestLifetimeMilliseconds(3000);

      OnData onData = [&savedMeta](const ptr_lib::shared_ptr<const Interest>& interest,
        const ptr_lib::shared_ptr<Data>& data)
        {
          savedMeta = data;
          cout << " > SAVED " << data->getName() << " " << data->getContent().toRawStr() << endl;
        };

      cout << "REQUEST " << i.getName() << endl;
      face.expressInterest(i, onData);
    };

    int metaFetchInterval = 500;
    int64_t now = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
    int64_t timestampCheck = now;

    while (true) {
      now = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
      if (now - timestampCheck >= metaFetchInterval)
      {
        timestampCheck = now;
        fetchMeta();
      }

      face.processEvents();
      usleep(10);
    }
}
